wallet/wallet.c

```diff
@@ -688,83 +657,99 @@ bool wallet_channel_save(struct wallet *w, struct wallet_channel *chan){
 	ok &= wallet_channel_config_save(w, &p->our_config);
 
 	/* Now do the real update */
-	ok &= db_exec(__func__, w->db, "UPDATE channels SET"
-		      "  shachain_remote_id=%"PRIu64","
-		      "  short_channel_id=%s,"
-		      "  state=%d,"
-		      "  funder=%d,"
-		      "  channel_flags=%d,"
-		      "  minimum_depth=%d,"
-		      "  next_index_local=%"PRIu64","
-		      "  next_index_remote=%"PRIu64","
-		      "  next_htlc_id=%"PRIu64","
-		      "  funding_tx_id=%s,"
-		      "  funding_tx_outnum=%d,"
-		      "  funding_satoshi=%"PRIu64","
-		      "  funding_locked_remote=%d,"
-		      "  push_msatoshi=%"PRIu64","
-		      "  msatoshi_local=%s,"
-		      "  shutdown_scriptpubkey_remote='%s',"
-		      "  shutdown_keyidx_local=%"PRId64","
-		      "  channel_config_local=%"PRIu64","
-		      "  last_tx=%s, last_sig=%s"
-		      " WHERE id=%"PRIu64,
-		      p->their_shachain.id,
-		      p->scid?tal_fmt(tmpctx,"'%s'", short_channel_id_to_str(tmpctx, p->scid)):"null",
-		      p->state,
-		      p->funder,
-		      p->channel_flags,
-		      p->minimum_depth,
-		      p->next_index[LOCAL],
-		      p->next_index[REMOTE],
-		      p->next_htlc_id,
-		      p->funding_txid?tal_fmt(tmpctx, "'%s'", tal_hexstr(tmpctx, p->funding_txid, sizeof(struct sha256_double))):"null",
-		      p->funding_outnum,
-		      p->funding_satoshi,
-		      p->remote_funding_locked,
-		      p->push_msat,
-		      p->our_msatoshi?tal_fmt(tmpctx, "%"PRIu64, *p->our_msatoshi):"NULL",
-		      p->remote_shutdown_scriptpubkey?tal_hex(tmpctx, p->remote_shutdown_scriptpubkey):"",
-		      p->local_shutdown_idx,
-		      p->our_config.id,
-		      db_serialize_tx(tmpctx, p->last_tx),
-		      db_serialize_signature(tmpctx, p->last_sig),
-		      chan->id);
+	stmt = db_prepare(w->db, tal_fmt(w, "UPDATE channels SET"
+					 "  shachain_remote_id=?,"
+					 "  short_channel_id=?,"
+					 "  state=?,"
+					 "  funder=?,"
+					 "  channel_flags=?,"
+					 "  minimum_depth=?,"
+					 "  next_index_local=?,"
+					 "  next_index_remote=?,"
+					 "  next_htlc_id=?,"
+					 "  funding_tx_id=?,"
+					 "  funding_tx_outnum=?,"
+					 "  funding_satoshi=?,"
+					 "  funding_locked_remote=?,"
+					 "  push_msatoshi=?,"
+					 "  msatoshi_local=?,"
+					 "  shutdown_scriptpubkey_remote=?,"
+					 "  shutdown_keyidx_local=?,"
+					 "  channel_config_local=?,"
+					 "  last_tx=?, last_sig=?"
+					 " WHERE id=?"));
+	sqlite3_bind_int64(stmt, 1, p->their_shachain.id);
+	if (p->scid)
+		sqlite3_bind_short_channel_id(stmt, 2, p->scid);
```

Comment:
That's annoying.  I'd prefer it to fail on unbound parameters, and force us to use sqlite3_bind_null().  If we miss one, we'd not get an error :(

Reply:
We could write a scanf()-style wrapper which did this for us.  But it seems overkill (cool idea for a ccan module though!)